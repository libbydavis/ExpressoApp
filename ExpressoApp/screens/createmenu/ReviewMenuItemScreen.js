import React, {Component, useState} from "react";
import {Text, View, StyleSheet, Image, TouchableOpacity, TouchableWithoutFeedback, TextInput} from "react-native";
import BouncyCheckbox from 'react-native-bouncy-checkbox';
import QuantityInput from "../addmenuitem/QuantityInput";
import AddToCartButton from "../cart/AddToCartButton";

//TODO: options lists

const ReviewMenuItemScreen = ({navigation, route}) => {
    const UNIT_PRICE = route.params.price;
    const [cartItem, setCartItem] = useState({
        title: route.params.title,
        price: UNIT_PRICE,
        currentPrice: route.params.price,
        quantity: 1,
        notes: '',
        options: ''
    })

    const receiveQuantity = (value) => {
        setCartItem({...this.state, ['quantity']: value});
        let newPrice = UNIT_PRICE * value;
        setCartItem({...cartItem, ['currentPrice']: newPrice});
    }

    const setNotes = (notes) => {
        setCartItem({...cartItem, ['notes']: notes});
    }

    const toggleOptions = (item, title, isChecked) => {
        let optionsCopy = [...cartItem.options];
        if (isChecked) {
            optionsCopy = [...optionsCopy, item];
            setCartItem({...cartItem, ['options']: optionsCopy});
        }
        else {
            let index = cartItem.options.indexOf(item);
            optionsCopy.splice(index, 1);
            setCartItem({...cartItem, ['options']: optionsCopy});
        }
    }

    return (
        <View>
            <TouchableWithoutFeedback onPress={() => navigation.navigate('AddMenuItem')}>
                <Text style={styles.backButtonText}>Back To Menu</Text>
            </TouchableWithoutFeedback>
            <View style={styles.pageView}>
                <View style={styles.contentView}>
                    <Image source={{uri: route.params.image}}></Image>
                    <Text style={styles.title}>{cartItem.title}</Text>
                    <Text style={styles.description}>{route.params.description}</Text>
                    <QuantityInput receiveValue={receiveQuantity} initialQuantity={cartItem.quantity}></QuantityInput>

                    <View style={styles.checkListView}>
                        {
                            route.params.optionLists.map((item, index) => {
                                return (
                                    <View key={index} style={styles.checklist}>
                                        <Text style={styles.checkListTitle}>{item.title}</Text>
                                        {
                                            item.options.map((innerItem, innerIndex) => {
                                                return <BouncyCheckbox iconStyle={{borderColor: '#25a2af'}}
                                                                       fillColor={'#25a2af'}
                                                                       text={innerItem}
                                                                       style={styles.checkItem}
                                                                       key={innerIndex}
                                                                       textStyle={styles.checkboxText}
                                                                       onPress={(isChecked) => toggleOptions(innerItem, item.title, isChecked)}/>
                                            })
                                        }
                                    </View>
                                )

                            })
                        }
                    </View>
                    <Text style={styles.price}>$ {cartItem.currentPrice}</Text>
                    <TextInput onChangeText={(text) => setNotes(text)} placeholder={"additional notes"} style={styles.notesBox} multiline={true}/>
                    <AddToCartButton item={cartItem}></AddToCartButton>
                </View>
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    pageView: {
        alignItems: 'center',
        marginTop: 15
    },
    contentView: {
        justifyContent: 'flex-start'
    },
    title: {
        fontFamily: 'Monserrat-Bold',
        fontSize: 25,
        marginBottom: 10,
        marginTop: 15
    },
    description: {
        fontFamily: 'Monserrat-Regular',
        marginBottom: 30,
    },
    price: {
        fontFamily: 'Monserrat-Bold',
        fontSize: 16,
        marginBottom: 20,
    },
    backButtonText: {
        fontFamily: 'Monserrat-ExtraBold',
        marginLeft: 22,
        fontSize: 15
    },
    checkListView: {
        marginBottom: 20,
        flexDirection: 'row'
    },
    checkItem: {
        marginBottom: 3,
        marginTop: 3,
    },
    notesBox: {
        borderBottomWidth: 1,
        borderStartWidth: 1,
        borderLeftWidth: 1,
        borderRightWidth: 1,
        borderTopWidth: 1,
        paddingLeft: 20,
        paddingRight: 20,
        marginBottom: 15
    },
    checkListTitle: {
        marginBottom: 5,
        fontSize: 16
    },
    checklist: {
        marginRight: 25
    },
    checkboxText: {
        color: 'black',
        textDecorationLine: 'none'
    }
})

export default ReviewMenuItemScreen;
