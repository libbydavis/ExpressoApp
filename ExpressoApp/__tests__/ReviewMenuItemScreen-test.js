// __tests__/ReviewMenuItemScreen-test.js
import React from 'react';
import renderer from 'react-test-renderer';
import ReviewMenuItemScreen from '../screens/createmenu/ReviewMenuItemScreen';
import {shallow} from "react-native/jest/renderer";


test('renders correctly', () => {
    const reviewMenuItem = renderer.create(<ReviewMenuItemScreen route={{params : {title: 'hotdog', price: 10, description: 'tasty hotdog', optionLists: []}}}/>).toJSON();
    expect(reviewMenuItem).toMatchSnapshot();
});

test('state changes correctly', () => {
    const wrapper = shallow(<ReviewMenuItemScreen props={{navigation: '', route: {params : {title: 'hotdog', price: 10, description: 'tasty hotdog', optionLists: []}}}}></ReviewMenuItemScreen>)
    //maybe use button to navigate to reviewmenuitemscreen to simulate click?
    expect(wrapper.state).toEqual({ title: 'hotdog', price: 10, currentPrice: 10, quantity: 1, notes: '', options: ''});

});
