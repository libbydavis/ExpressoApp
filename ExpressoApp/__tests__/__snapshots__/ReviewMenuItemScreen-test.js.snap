// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "marginBottom": 15,
          "marginTop": 8,
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/ExpressoLogo.png",
          }
        }
        style={
          Object {
            "height": 50,
            "width": 200,
          }
        }
      />
    </View>
  </View>
  <Text
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "fontFamily": "Monserrat-ExtraBold",
        "fontSize": 15,
        "marginLeft": 22,
      }
    }
  >
    Back To Menu
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginTop": 15,
      }
    }
  >
    <View
      style={
        Object {
          "justifyContent": "flex-start",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": undefined,
          }
        }
      />
      <Text
        style={
          Object {
            "fontFamily": "Monserrat-Bold",
            "fontSize": 25,
            "marginBottom": 10,
            "marginTop": 15,
          }
        }
      >
        hotdog
      </Text>
      <Text
        style={
          Object {
            "fontFamily": "Monserrat-Regular",
            "marginBottom": 30,
          }
        }
      >
        tasty hotdog
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/minusButton.png",
              }
            }
            style={
              Object {
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          rejectResponderTermination={true}
          style={
            Object {
              "marginLeft": 10,
              "textAlignVertical": "top",
            }
          }
          underlineColorAndroid="transparent"
          value="1"
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/addButton.png",
              }
            }
            style={
              Object {
                "height": 25,
                "width": 25,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginBottom": 20,
          }
        }
      />
      <Text
        style={
          Object {
            "fontFamily": "Monserrat-Bold",
            "fontSize": 16,
            "marginBottom": 20,
          }
        }
      >
        $ 
        10
      </Text>
      <TextInput
        allowFontScaling={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="additional notes"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderStartWidth": 1,
            "borderTopWidth": 1,
            "marginBottom": 15,
            "paddingLeft": 20,
            "paddingRight": 20,
          }
        }
        underlineColorAndroid="transparent"
      />
      <View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#25a2af",
              "borderRadius": 10,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#ffffff",
                "padding": 10,
              }
            }
          >
            Add To Cart
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
